napari/layers/labels/labels.py

```diff
@@ -1475,6 +1451,34 @@ def data_setitem(self, indices, value, refresh=True):
         # update the labels image
         self.data[indices] = value
 
+        if not (
+            isinstance(self.data, np.ndarray)  # numpy array, or
+            or hasattr(self.data, 'data')
+            and type(self.data.data) is np.ndarray  # numpy-backed xarray
+        ):
+            # In the absence of slicing, the current slice becomes
+            # invalidated by data_setitem; only in the special case of a NumPy
+            # array, or a NumPy-array-backed Xarray, is the slice a view and
+            # therefore updated automatically.
+            # For other types, we update it manually here.
+            dims = self._slice.dims
+            point = np.round(self.world_to_data(dims.point)).astype(int)
+            point_not_displayed = [point[i] for i in dims.not_displayed]
+            indices_not_displayed = [indices[i] for i in dims.not_displayed]
+            index_in_slice = np.logical_and.reduce(
+                [
+                    index == pt
+                    for index, pt in zip(
+                        indices_not_displayed, point_not_displayed
+                    )
+                ],
+                axis=0,
+            )
+            displayed_indices = tuple(
+                indices[i][index_in_slice] for i in dims.displayed
+            )
+            self._slice.image.raw[displayed_indices] = value
```

Comment:
Only updating the raw image slice has two possible issues.

1\. The thumbnail is not updated.
 
In general, the thumbnail may have a different shape so we can't just reuse `displayed_indices` and do something similar unfortunately.

2\. The image slice view is not updated.

This doesn't matter for the rendered canvas result since we do the mapping in `_partial_labels_refresh` and pass it through to vispy through the `labels_update` event. I think the only time it's used is in `_ImageBase._calc_data_range` and you're definitely in a better position to judge whether that matters for labels colormapping!

If needed, I think you could fix this fairly easily by doing something like `self._slice.image.view[displayed_indices] = self._raw_to_displayed(self._slice.image.raw[displayed_indices])` (or maybe just `:` instead of `displayed_indices` depending on how `_raw_to_displayed` works). Either here or in `_partial_labels_refresh`.

Both these issues exist on `main` (plus the much bigger one this PR fixes), so I don't think either are blocking. But we may at least want an issue to follow up on.

Reply:
> I think the only time it's used is in _ImageBase._calc_data_range and you're definitely in a better position to judge whether that matters for labels colormapping!

Not one bit. 😃 

Should we perhaps deprecate `.view`?

Reply:
That would be lovely. As it's a member of a private property, I'd prefer just removing it so we can simplify code in a couple of places.

Reply:
Should I create an issue to track this? I don't expect the deprecation/removal of that property as part of this PR.

Reply:
I think that's a good idea. Thank you!

Reply:
Done: https://github.com/napari/napari/issues/6185